### PR TITLE
resource/aws_acm_certificate_validation: Prevent crash on validation_record_fqdns

### LINKED ISSUE #3330: Terraform apply crashed when configuring aws_acm_certificate_validation
_This issue was originally opened by @uccmen as hashicorp/terraform#17320. It was migrated here as a result of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

### Terraform Version

```
v0.11.3

```

### Terraform Configuration Files

```hcl
resource "aws_acm_certificate" "app-production" {
  domain_name = "app.example.io"
  validation_method = "EMAIL"
}

resource "aws_acm_certificate_validation" "app-production" {
  certificate_arn = "${aws_acm_certificate.app-production.arn}"
  validation_record_fqdns = ["example.io"]
}
```

### Crash Output

https://gist.github.com/uccmen/16c3ba8ce677e99f6232052937d53357

### Expected Behavior

Terraform plan should complete successfully

### Actual Behavior

Terraform plan crashed

### Steps to Reproduce

1. `terraform init`
2. `terraform plan`
3. `terraform apply`

### Additional Context

1. Terraform is running in a CI system (Jenkins)
2. Image used is `light` i.e. `hashicorp/terraform:light`
3. My editor (IntelliJ IDEA Ultimate) wasn't able to pick up `arn` param in `${aws_acm_certificate.app-production.arn}`
4. Both domain names and fqdns provided above are examples

### References

https://docs.aws.amazon.com/acm/latest/userguide/gs-acm-validate-email.html
https://docs.aws.amazon.com/acm/latest/userguide/setup-email.html
https://www.terraform.io/docs/providers/aws/r/acm_certificate_validation.html

### PR DESCRIPTION
**Pull request by bflad**
Fixes #3330 

Other changes:
* Will retry until domain validation options are actually available when using validation_record_fqdns
* Will now return an error if validation_record_fqdns is used on EMAIL validation method ACM certificates (with associated acceptance test)
* Add docs example for email validation
* Fix docs link to aws_acm_certificate resource instead of self-referencing

```
make testacc TEST=./aws TESTARGS='-run=TestAccAwsAcmCertificateValidation_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAwsAcmCertificateValidation_basic -timeout 120m
=== RUN   TestAccAwsAcmCertificateValidation_basic
--- PASS: TestAccAwsAcmCertificateValidation_basic (225.96s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	226.000s
```

### CODE REVIEW

**Comment by cemo** [comment]
@bflad Have you seen the comment https://github.com/terraform-providers/terraform-provider-aws/pull/2813#issuecomment-364701215. This is pretty much limiting usage. Is it possible to fix this  part as well?

File: aws/resource_aws_acm_certificate_validation.go [hunk 1]
```diff
@@ -83,11 +84,38 @@ func resourceAwsAcmCertificateValidationCreate(d *schema.ResourceData, meta inte
...
-func resourceAwsAcmCertificateCheckValidationRecords(validation_record_fqdns []interface{}, cert *acm.CertificateDetail) error {
+func resourceAwsAcmCertificateCheckValidationRecords(validation_record_fqdns []interface{}, cert *acm.CertificateDetail, conn *acm.ACM) error {
 	expected_fqdns := make([]string, len(cert.DomainValidationOptions))
+
+	if len(cert.DomainValidationOptions) == 0 {
+		input := &acm.DescribeCertificateInput{
+			CertificateArn: cert.CertificateArn,
+		}
+		err := resource.Retry(1*time.Minute, func() *resource.RetryError {
+			log.Printf("[DEBUG] Certificate domain validation options empty, retrying")
```
Comment (radeksimko) [inline aws/resource_aws_acm_certificate_validation.go]: Nitpick: Having the cert ARN in the log message would make it IMO more helpful.

File: aws/resource_aws_acm_certificate_validation.go [hunk 2]
```diff
@@ -83,11 +84,38 @@ func resourceAwsAcmCertificateValidationCreate(d *schema.ResourceData, meta inte
...
+			output, err := conn.DescribeCertificate(input)
+			if err != nil {
+				return resource.NonRetryableError(err)
+			}
+			if len(output.Certificate.DomainValidationOptions) == 0 {
+				return resource.RetryableError(errors.New("Certificate domain validation options empty"))
```
Comment (radeksimko) [inline aws/resource_aws_acm_certificate_validation.go]: Nitpick: Likewise ^

File: aws/resource_aws_acm_certificate_validation.go (hunk 1 above)
Comment (bflad) [reply aws/resource_aws_acm_certificate_validation.go]: Oh good catch! Will add

File: aws/resource_aws_acm_certificate_validation.go (hunk 2 above)
Comment (bflad) [reply aws/resource_aws_acm_certificate_validation.go]: 👍

**Comment by bflad** [comment]
> I have mixed feelings about supporting email confirmations in general. I feel like it's out of the usual/expected scope of Terraform, because email confirmation is rarely well automated and letting user wait until someone clicks a link is odd.

Me too and it certainly is 😂  but I imagine some (unfortunate) environments are stuck with email validation for whatever reason, so even if its janky, it provides them a starting path into Terraform usage. There's no technical reason to prevent it as the support on our end was simply allowing the additional validation method to pass through to the ACM API in the certificate resource and prevent email validation from working with this DNS-specific attribute in the validation resource. Email validation can timeout just the same as if ACM's DNS validation system has issues.

**Comment by bflad** [comment]
This has been released in [version 1.10.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1100-february-24-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Pull request merged**